packages/embark/src/cmd/cmd_controller.js

```diff
@@ -210,25 +213,31 @@ class EmbarkController {
           engine.events.emit("status", __("Ready").green);
         });
 
-        engine.events.on('file-event', async ({ fileType, path }) => {
-          // TODO: re-add async.cargo / or use rxjs to use latest request in the queue
-          console.dir("-- before timeout - file changed")
-
-          if (fileType === 'contract' || fileType === 'config') {
-            let contractsFiles = await engine.events.request2("config:contractsFiles");
-            let compiledContracts = await engine.events.request2("compiler:contracts:compile", contractsFiles);
-            let _contractsConfig = await engine.events.request2("config:contractsConfig");
-            let contractsConfig = cloneDeep(_contractsConfig);
-            let [contractsList, contractDependencies] = await engine.events.request2("contracts:build", contractsConfig, compiledContracts);
-            await engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies);
-          } else if (fileType === 'asset') {
-            engine.events.request('pipeline:generateAll', () => {
-              console.dir("outputDone")
-              engine.events.emit('outputDone');
-            });
-          }
+        const fileChanges$ = Observable.create(obs => {
+          engine.events.on('file-event', (data) => {
+            obs.next(data);
+          });
         });
 
+        fileChanges$.pipe(
+          exhaustMapWithLastIgnored(({ fileType, path }) => {
+            if (fileType === 'contract' || fileType === 'config') {
+              const contractFiles$ = from(engine.events.request2("config:contractsFiles"));
+              const compiledContracts$ = contractFiles$.pipe(
+                mergeMap(contractsFiles => from(engine.events.request2("compiler:contracts:compile", contractsFiles)))
+              );
+              const contractsConfig$ = from(engine.events.request2("config:contractsConfig"));
+
+              return forkJoin(contractsConfig$, compiledContracts$).pipe(
+                mergeMap(([contractsConfig, compiledContracts]) => from(engine.events.request2("contracts:build", contractsConfig, compiledContracts))),
+                mergeMap(([contractsList, contractDependencies]) => from(engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies)))
+              );
```

Comment:
Few notes here:

- `from()` turns any thing that can be turned into an Observable into an Observable, in this case, Promises
- We can't use `async/await` here because it's crucial that the callback of `exhaustMapWithLastIgnored()` returns an Observable that notifies the outer stream. Using `async` would make this callback automatically return a promise.
- `forkJoin()` is pretty much the equivalent of `Promise.all()`
- A lot of this boilerplate can go away if `engine.events.request()` would return an observable already (instead of a Promise)

Reply:
Also this entire logic could be extracted into another function for readability's sake, so we'd end up with:

```
fileChanges$.pipe(
  exhaustMapWithLastIgnored((fileType, path} => {
    if (fileType === 'contract || fileType === 'config') {
      return compileAndDeploy();
    } else if (fileType === 'asset') {
      return generateAssets();
    }
  })
).subscribe();
```